android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt

```diff
@@ -315,3 +324,84 @@ fun String.extractLogicalIdUuid() = this.substringAfter("/").substringBefore("/"
 fun Resource.addTags(tags: List<Coding>) {
   tags.forEach { this.meta.addTag(it) }
 }
+
+/**
+ * You provide a suspended function in Kotlin, which updates the due date of a task's dependent
+ * tasks based on the date of a related immunization. The function takes a [defaultRepository]
+ * parameter that is an instance of [DefaultRepository]. It then loops through all the tasks that
+ * this task is a part of, loads the dependent tasks and their related immunization resources from
+ * the repository, and updates the start date of the dependent task if it's scheduled to start
+ * before the immunization date plus the required number of days.
+ *
+ * We may potentially extend this function to consider the attributes of resources other than
+ * immunizations.
+ *
+ * @param defaultRepository An instance of DefaultRepository
+ */
+suspend fun Task.updateDependentTaskDueDate(defaultRepository: DefaultRepository): Task {
+  return apply {
+    if (hasPartOf()) {
+      partOf.forEach { task ->
+        val dependentTask =
+          defaultRepository.loadResource<Task>(task.reference.extractLogicalIdUuid())
+        if (dependentTask != null &&
+            dependentTask.hasOutput() &&
+            dependentTask.executionPeriod.hasStart() &&
+            dependentTask.hasInput() &&
+            (dependentTask.isDue() || dependentTask.isOverDue() || dependentTask.isUpcoming())
+        ) {
+          dependentTask.output?.forEach { dependentTaskOutputValue ->
+            if (dependentTaskOutputValue.hasValue()) {
+              val dependentTaskReference =
+                Reference(
+                  Json.decodeFromString<JsonObject>(dependentTaskOutputValue.value.toString())[
+                      REFERENCE]
+                    ?.jsonPrimitive
+                    ?.content
+                )
+              val encounterResource =
+                defaultRepository.loadResource<Encounter>(
+                  dependentTaskReference.reference.extractLogicalIdUuid()
+                )
+              encounterResource?.partOf?.reference?.let { partOfReference ->
+                try {
+                  val immunizationResource =
+                    defaultRepository.loadResource<Immunization>(
+                      partOfReference.extractLogicalIdUuid()
+                    )
```

Comment:
added the `try` because we can't guarantee what exists at the reference -- not sure if just a log and return is appropriate, but for now seems OK